### ISSUE TITLE
Zooming in or out after the map is rotated doesn't work correctly

### ISSUE THREAD

**Issue by heyajohnny** [Raruto/leaflet-rotate#19]
This only applies to touch devices. On desktop, it works as expected.

After the map is rotated, the zooming on touch devices makes sort of a jump. It looks like it's wanting to zoom in/out from the non rotated location. 
You can see the strange behavior in the [demo](https://raruto.github.io/leaflet-rotate/examples/leaflet-rotate.html) on a mobile device.

Do you have any idea how to fix this, or lead me to the right direction to fix it myself.

**Comment by Raruto** [comment]
Hi @heyajohnny,

> You can see the strange behavior in the [demo](https://raruto.github.io/leaflet-rotate/examples/leaflet-rotate.html) on a mobile device.

I can't test this because I don't have a touch device, however if you can upload a gif it could be useful for others looking for a solution too.

> Do you have any idea how to fix this, or lead me to the right direction to fix it myself.

Here are some classes you can start looking on:

- [src/map/handler/TouchGestures.js](https://github.com/Raruto/leaflet-rotate/blob/master/src/map/handler/TouchGestures.js)
- [src/map/handler/TouchRotate.js](https://github.com/Raruto/leaflet-rotate/blob/master/src/map/handler/TouchRotate.js)
- [src/map/handler/TouchZoom.js](https://github.com/Raruto/leaflet-rotate/blob/master/src/map/handler/TouchZoom.js)

If you don't know what they are, in the official documentation you can find some information about them: [Extending Leaflet, New Handlers and Controls](https://leafletjs.com/examples/extending/extending-3-controls.html)

If it can help for debugging, here is the library used for this purpose in the above [demo](https://raruto.github.io/leaflet-rotate/examples/leaflet-rotate.html):

- [lib/debug.js](https://github.com/Raruto/leaflet-rotate/blob/master/lib/debug.js)


_Have a nice day,
Raruto_

**Comment by heyajohnny** [comment]
This is the bug: 
https://user-images.githubusercontent.com/20553716/213726439-f623430e-6e09-41aa-914e-24d615aeef35.mp4
It seems that the first time zooming works perfectly, but the upfollowing times, it just ignores the extended touchXXX functions.

I changed the following lines in [src/map/handler/TouchGestures.js](https://github.com/Raruto/leaflet-rotate/blob/master/src/map/handler/TouchGestures.js)
![image](https://user-images.githubusercontent.com/20553716/213728662-ee9e10af-3cab-4b29-8d30-b1ebb5fab19c.png)
(Changing the DOM event listeners from 'document' to 'window' seems to fix the issue):
https://user-images.githubusercontent.com/20553716/213727618-6425d2af-f4ee-499b-b88c-cdc243bf6c74.mp4

**Comment by Raruto** [comment]
> BEFORE: https://user-images.githubusercontent.com/20553716/213726439-f623430e-6e09-41aa-914e-24d615aeef35.mp4
> AFTER: https://user-images.githubusercontent.com/20553716/213727618-6425d2af-f4ee-499b-b88c-cdc243bf6c74.mp4

I can definitely see badly, but I struggle to notice the difference between the two videos..

>  It seems that the first time zooming works perfectly, but the upfollowing times, it just ignores the extended touchXXX functions. I changed the following lines in [src/map/handler/TouchGestures.js](https://github.com/Raruto/leaflet-rotate/blob/master/src/map/handler/TouchGestures.js)
> ...
> (Changing the DOM event listeners from 'document' to 'window' seems to fix the issue)
>
> Can you please update the code or accept my pull request? Thnx!

Probably the latest leaflet updates (> v1.7) there were several that potentially destroyed this override, to do something right we would have to check every single file of this plugin with what is present in the core to be sure that these classes / file still exist (including its properties and methods).

For example, an equivalent core library handler (`Map.TouchZoom.js` v1.9.3) appears to make use of another additional DOM event (ref: `touchcancel`):

```
DomEvent.off(document, 'touchend touchcancel', this._onTouchEnd, this);
```

BTW, the core also seems to listen on the `document` object (maybe ambitious, but could it also not be just a bug specific to this plugin?)

https://github.com/Leaflet/Leaflet/blob/5c29d615563835d373a51b91fc3513a17907da96/src/map/handler/Map.TouchZoom.js#L60-L61

https://github.com/Leaflet/Leaflet/blob/5c29d615563835d373a51b91fc3513a17907da96/src/map/handler/Map.TouchZoom.js#L114-L115

**Comment by Raruto** [comment]
@heyajohnny taking a quick look at the following handler:

https://github.com/Raruto/leaflet-rotate/blob/6bfcccecf50143fd927e5f347a3f4d5a930c2f56/src/map/handler/TouchGestures.js#L1-L3

It's possible that they're already conflicting, have you tried doing this by any chance?

```js
// ref: https://leafletjs.com/reference#map-touchzoom

L.map('map', { touchZoom: false, touchGestures: true })
```

or 

```js
map.touchZoom.disable()
map.touchGestures.enable()
```

**Comment by heyajohnny** [comment]
@Raruto Thnx for all the help! Yeah, it was a quick dirty fix and I wasn't really familiar with the leaflet map. 

> It's possible that they're already conflicting, have you tried doing something like this?
No, I tried enabling / disabling these handlers, but that is not the problem. They work as expected.

The problem is in the 'TouchGestures.js'

I've compared and debugged the 'TouchGestures.js' with the current leaflet [TouchZoom](https://github.com/Leaflet/Leaflet/blob/main/src/map/handler/Map.TouchZoom.js)
https://github.com/Leaflet/Leaflet/blob/5c29d615563835d373a51b91fc3513a17907da96/src/map/handler/Map.TouchZoom.js
I made some changes to the 'TouchGestures.js'.
[Here](https://github.com/heyajohnny/leaflet-rotate/blob/master/src/map/handler/TouchGestures.js) is my updated 'TouchGestures.js'
Maybe you could compare and add the fix to your new [Pull request](https://github.com/Raruto/leaflet-rotate/pull/21)
Or maybe I can help with it. Please let me know if I could help in any way?


https://user-images.githubusercontent.com/20553716/214035278-e882af0a-9761-42c3-8e68-ef3a06baf4a0.mp4

**Comment by Raruto** [comment]
> Or maybe I can help with it. Please let me know if I could help in any way?

Oh well, in order to add support to leaflet@v1.9.3 it would also be nice to figure out how to rewrite the following functions in order to make it work correctly in a rotated "world".

## • [src/map/Map.js::panInsideBounds(bounds, options)](https://github.com/Leaflet/Leaflet/blob/c8116bac09205b35cb91ca0234b36ff90c0a6b4c/src/map/Map.js#L497-L510)

```js
// @method panInsideBounds(bounds: LatLngBounds, options?: Pan options): this
// Pans the map to the closest view that would lie inside the given bounds (if it's not already), controlling the animation using the options specific, if any.
panInsideBounds(bounds, options) {
  this._enforcingBounds = true;
  const center = this.getCenter(),
  newCenter = this._limitCenter(center, this._zoom, toLatLngBounds(bounds));

  if (!center.equals(newCenter)) {
    this.panTo(newCenter, options);
  }

  this._enforcingBounds = false;
  return this;
},
```

## • [src/map/Map.js#L512-L538](https://github.com/Leaflet/Leaflet/blob/c8116bac09205b35cb91ca0234b36ff90c0a6b4c/src/map/Map.js#L512-L538)

```js
panInside(latlng, options) {
  options = options || {};

  const paddingTL = toPoint(options.paddingTopLeft || options.padding || [0, 0]),
  paddingBR = toPoint(options.paddingBottomRight || options.padding || [0, 0]),
  pixelCenter = this.project(this.getCenter()),
  pixelPoint = this.project(latlng),
  pixelBounds = this.getPixelBounds(),
  paddedBounds = toBounds([pixelBounds.min.add(paddingTL), pixelBounds.max.subtract(paddingBR)]),
  paddedSize = paddedBounds.getSize();

  if (!paddedBounds.contains(pixelPoint)) {
    this._enforcingBounds = true;
    const centerOffset = pixelPoint.subtract(paddedBounds.getCenter());
    const offset = paddedBounds.extend(pixelPoint).getSize().subtract(paddedSize);
    pixelCenter.x += centerOffset.x < 0 ? -offset.x : offset.x;
    pixelCenter.y += centerOffset.y < 0 ? -offset.y : offset.y;
    this.panTo(this.unproject(pixelCenter), options);
    this._enforcingBounds = false;
  }
  return this;
},
```

As Ivan pointed out before, many of the bugs that can be found within the application (eg. [fnicollet/Leaflet/issues](https://github.com/fnicollet/Leaflet/issues)) are also due to an incorrect initial reference for the calculated bounds (ie. the displayed portion of the map).

Here for example: https://github.com/Raruto/leaflet-rotate/issues/18#issuecomment-1399349255, a related bug that has become more apparent in leaflet@v1.8.0 (but even before it would have given problems if you ever tried to write a similar feature).

**Comment by Raruto** [comment]
Patch released in **v0.2.0**
